src/batch/src/executor/join/hash_join.rs

```diff
@@ -12,319 +12,1403 @@
+            yield chunk?.reorder_columns(&self.output_indices)
+        }
     }
 
-    #[inline(always)]
-    pub(super) fn batch_size(&self) -> usize {
-        self.batch_size
+    #[try_stream(boxed, ok = DataChunk, error = RwError)]
+    async fn do_inner_join(
+        EquiJoinParams {
+            probe_side,
+            probe_key_idxs,
+            build_side,
+            full_data_types,
+            hash_map,
+            next_build_row_with_same_key,
+            ..
+        }: EquiJoinParams<K>,
+    ) {
+        let mut chunk_builder = DataChunkBuilder::with_default_size(full_data_types);
+        #[for_await]
+        for probe_chunk in probe_side.execute() {
+            let probe_chunk = probe_chunk?;
+            let probe_keys = K::build(&probe_key_idxs, &probe_chunk)?;
+            for (probe_row_id, probe_key) in probe_keys.iter().enumerate() {
+                for build_row_id in
+                    next_build_row_with_same_key.row_id_iter(hash_map.get(probe_key).copied())
+                {
+                    let build_chunk = &build_side[build_row_id.chunk_id()];
+                    if let Some(spilled) = Self::append_one_row(
+                        &mut chunk_builder,
+                        &probe_chunk,
+                        probe_row_id,
+                        build_chunk,
+                        build_row_id.row_id(),
+                    )? {
+                        yield spilled
+                    }
+                }
+            }
+        }
+        if let Some(spilled) = chunk_builder.consume_all()? {
+            yield spilled
+        }
     }
 
-    #[inline(always)]
-    pub(super) fn has_non_equi_cond(&self) -> bool {
-        self.cond.is_some()
+    #[try_stream(boxed, ok = DataChunk, error = RwError)]
+    async fn do_inner_join_with_non_equi_condition(
+        params: EquiJoinParams<K>,
+        cond: BoxedExpression,
+    ) {
+        #[for_await]
+        for chunk in Self::do_inner_join(params) {
+            let mut chunk = chunk?;
+            chunk.set_visibility(cond.eval(&chunk)?.as_bool().iter().collect());
+            if chunk.cardinality() > 0 {
+                yield chunk.compact()?
+            }
+        }
     }
 
-    #[allow(clippy::too_many_arguments)]
-    pub fn new(
-        join_type: JoinType,
-        left_key_columns: Vec<usize>,
-        left_key_types: Vec<DataType>,
-        left_col_len: usize,
-        right_key_columns: Vec<usize>,
-        right_key_types: Vec<DataType>,
-        right_col_len: usize,
-        full_data_types: Vec<DataType>,
-        batch_size: usize,
-        cond: Option<BoxedExpression>,
-    ) -> Self {
-        Self {
-            join_type,
-            left_key_columns,
-            left_key_types,
-            left_col_len,
-            right_key_columns,
-            right_key_types,
-            right_col_len,
+    #[try_stream(boxed, ok = DataChunk, error = RwError)]
+    async fn do_left_outer_join(
```

Comment:
I remember that we treat leaft side as build side, and this implementation seems to treat it as probe side?